Refactor tooltip positioning with `useLayoutEffect` and portals.

H0taru-9F opened the pull request:
Replaced CSS-based tooltip positioning with JS-driven dynamic positioning for better placement flexibility. Integrated React portals for rendering tooltips at the `document.body` level, ensuring consistent visibility. Updated components to support these changes and improved type definitions for enhanced React compatibility.

H0taru-9F committed:
Refactor tooltip positioning with `useLayoutEffect` and portals.

Replaced CSS-based tooltip positioning with JS-driven dynamic positioning for better placement flexibility. Integrated React portals for rendering tooltips at the `document.body` level, ensuring consistent visibility. Updated components to support these changes and improved type definitions for enhanced React compatibility.

H0taru-9F committed:
Refactor tooltip positioning with `useLayoutEffect` and portals.

Replaced CSS-based tooltip positioning with JS-driven dynamic positioning for better placement flexibility. Integrated React portals for rendering tooltips at the `document.body` level, ensuring consistent visibility. Updated components to support these changes and improved type definitions for enhanced React compatibility.

H0taru-9F committed:
Refactor tooltip positioning with `useLayoutEffect` and portals.

Replaced CSS-based tooltip positioning with JS-driven dynamic positioning for better placement flexibility. Integrated React portals for rendering tooltips at the `document.body` level, ensuring consistent visibility. Updated components to support these changes and improved type definitions for enhanced React compatibility.

mhevyk reviewed (commented):
The sizes of icons are broken when we use truncation. Please take a look

![image](https://github.com/user-attachments/assets/1f04dfa0-fe5a-45a9-9768-819287aa697e)

mhevyk commented on src/styles/app-tooltip/AppTooltip.tsx:
```diff
@@ -35,6 +47,57 @@ const AppTooltip = forwardRef<HTMLDivElement, AppTooltipProps>(function (
+    });
+  }, [isVisible, position, tooltipId]);
+
+  const trigger = cloneElement(children, {
+    ref: triggerRef,
+  });
```
I would not recommend using clone element to pass ref, because it may confict with the existing ref which the element has. I propose to either consider using additional div wrapper for it or applying render prop react pattern to render tooltip element instead of using children, because children prop is difficult to configure

Reference: https://www.patterns.dev/react/render-props-pattern/

mhevyk commented on src/styles/app-tooltip/AppTooltip.tsx:
```diff
@@ -35,6 +47,57 @@ const AppTooltip = forwardRef<HTMLDivElement, AppTooltipProps>(function (
+    }
+    const triggerRect = triggerRef.current.getBoundingClientRect();
+
+    const tooltipElement = document.getElementById(`tooltip-${tooltipId}`);
```
Can we avoid searching through document? If not, we should at least search tooltipContainer.getElementById, not whole document. I also recommend changing the default place of rending portal, see how I did it with modal's portal

mhevyk reviewed (commented):
![image](https://github.com/user-attachments/assets/0f9e20fd-233c-479e-9192-2116cd362708)

I see class app-tooltip--undefined, fix it

mhevyk committed:
Rebase branch

mhevyk committed:
Rebase branch

H0taru-9F committed:
Merge remote-tracking branch 'origin/bugfix/315/app-select-long-label' into bugfix/315/app-select-long-label

H0taru-9F committed:
Refactor AppTooltip to use a render prop for flexibility

Replaces `children` with a `render` prop in AppTooltip for improved customization and type safety. Updates related components and stories to align with the new API. This change allows more precise control over the tooltip's trigger elements.

H0taru-9F committed:
Refactor AppTooltip to use a render prop for flexibility

Replaces `children` with a `render` prop in AppTooltip for improved customization and type safety. Updates related components and stories to align with the new API. This change allows more precise control over the tooltip's trigger elements.

H0taru-9F committed:
Refactor modal and tooltip components to use a reusable domPortal utility

H0taru-9F committed:
Refactor modal and tooltip components to use a reusable domPortal utility

H0taru-9F committed:
Set minimum width for icons to improve layout consistency

Added the `min-width: fit-content` rule to ensure icons maintain proper spacing and alignment. This change enhances visual consistency across the components.

H0taru-9F committed:
Set minimum width for icons to improve layout consistency

Added the `min-width: fit-content` rule to ensure icons maintain proper spacing and alignment. This change enhances visual consistency across the components.

H0taru-9F committed:
Refactor AppTooltip to improve type flexibility and consistency

Updated `AppTooltip` to better support generic types and added "sm" to allowed tooltip sizes. Refactored the component to simplify type usage, improve clarity, and maintain compatibility with `forwardRef`.

H0taru-9F committed:
Refactor AppTooltip to improve type flexibility and consistency

Updated `AppTooltip` to better support generic types and added "sm" to allowed tooltip sizes. Refactored the component to simplify type usage, improve clarity, and maintain compatibility with `forwardRef`.

mhevyk commented on src/styles/app-overflow-text/AppOverflowText.tsx:
```diff
@@ -38,10 +38,20 @@ const AppOverflowText = <Element extends ElementType = ElementType>({
-      <AppTooltip content={children} {...tooltipProps}>
-        {textElement}
-      </AppTooltip>
+      <AppTooltip
+        content={content}
+        render={({ ref }) => {
+          const divRef = ref as RefObject<HTMLDivElement>;
+          return <div ref={divRef}>{textElement}</div>;
+        }}
+        {...tooltipProps}
+      />
```
Kudos @H0taru-9F! But I'd use explicit generic, please take a look at this syntax:

```tsx
return (
      <AppTooltip<HTMLDivElement>
        content={content}
        render={({ ref }) => <div ref={ref}>{textElement}</div>}
        {...tooltipProps}
      />
```

H0taru-9F replied:
it looks cool

mhevyk commented on src/styles/app-tooltip/AppTooltip.stories.tsx:
```diff
@@ -34,67 +34,56 @@ const meta: Meta<AppTooltipProps> = {
 export default meta;
 
-type Story = StoryObj<AppTooltipProps>;
+type ButtonStory = StoryObj<AppTooltipProps<HTMLButtonElement>>;
+type InputStory = StoryObj<AppTooltipProps<HTMLInputElement>>;
+type LinkStory = StoryObj<AppTooltipProps<HTMLAnchorElement>>;
```
👏🏼

mhevyk commented on src/styles/app-tooltip/AppTooltip.tsx:
```diff
@@ -1,20 +1,40 @@
-  { position = "top", width, content, children, className, ...props },
-  ref
-) {
+const AppTooltip = <T extends HTMLElement = HTMLElement>({
+  position = "top",
+  width = "sm",
+  content,
+  render,
+  className,
+  ...props
```
Is formatting ok here?

mhevyk committed:
AppSelect: Fix placement of menu while resize (#322)

* Fix placement of menu while resize

* Add comment

H0taru-9F committed:
Refactor AppOverflowText to simplify AppTooltip rendering

Removed unnecessary type casting and streamlined the AppTooltip usage by directly applying the generic type. This cleanup improves code readability and reduces redundant logic.

H0taru-9F committed:
Refactor AppOverflowText to simplify AppTooltip rendering

Removed unnecessary type casting and streamlined the AppTooltip usage by directly applying the generic type. This cleanup improves code readability and reduces redundant logic.

H0taru-9F committed:
Merge remote-tracking branch 'origin/bugfix/315/app-select-long-label' into bugfix/315/app-select-long-label
